File: SwiftUI/EdgeStudio/Data/AdvancedSettingsApplier.swift
import DittoSwift
import Foundation

/// Minimal DQL execution surface, so the advanced-settings apply logic can be tested
/// with a recording fake instead of a live Ditto instance.
protocol DQLExecuting: Sendable {
    /// Runs `query` and returns each result item's value dictionary.
    @discardableResult
    func runDQL(_ query: String, arguments: [String: Any?]) async throws -> [[String: Any?]]
}

extension Ditto: DQLExecuting {
    @discardableResult
    func runDQL(_ query: String, arguments: [String: Any?]) async throws -> [[String: Any?]] {
        let result = arguments.isEmpty
            ? try await store.execute(query: query)
            : try await store.execute(query: query, arguments: arguments)
        return result.items.map(\.value)
    }
}

/// Applies a database's Advanced Configuration to an open Ditto instance.
///
/// Split out of `DittoManager` for two reasons: the statement order relative to
/// `sync.start()` is safety-critical and needs a unit test that doesn't require
/// credentials, and the two lists need **opposite failure policies**:
///
/// - **Startup settings** are tuning knobs → best-effort. One bad parameter name is
///   reported, not fatal.
/// - **Sync scopes** are a data-containment control → fail-closed. If they cannot be
///   applied and verified, the caller must not start sync, because "scope missing"
///   means data the user marked device-local would replicate.
struct AdvancedSettingsApplier: Sendable {
    let executor: any DQLExecuting

    enum ApplyError: LocalizedError {
        case scopeStatementFailed(collections: [String], underlying: String)
        case scopeVerificationMismatch(expected: [String: String], actual: [String: String])
        case invalidScopes(String)

        var errorDescription: String? {
            switch self {
            case let .scopeStatementFailed(collections, underlying):
                return "Could not apply collection sync scopes for \(collections.joined(separator: ", ")). " +
                    "The database was not opened, to avoid syncing data you marked device-local. (\(underlying))"
            case let .scopeVerificationMismatch(expected, actual):
                return "Collection sync scopes were not stored as requested " +
                    "(expected \(expected), found \(actual)). The database was not opened, " +
                    "to avoid syncing data you marked device-local."
            case let .invalidScopes(detail):
                return "Collection sync scopes are invalid: \(detail). The database was not opened."
            }
        }
    }

    // MARK: - Startup Settings (best-effort)

    /// Applies each startup setting independently. A failure is recorded and skipped
    /// rather than aborting: an unrecognized parameter name is a typo, not a reason to
    /// lock the user out of their database.
    func applyStartupSettings(_ settings: [StartupSetting]) async -> AdvancedApplyResult {
        var result = AdvancedApplyResult()
        guard !settings.isEmpty else { return result }

        // Whole-list validation, not per-row: `validateSetting(_:others: [])` cannot see
        // duplicates, and the row cap was only ever enforced in the editor — so a
        // QR-imported or hand-edited config could issue thousands of statements, or two
        // conflicting writes to one parameter, silently.
        let partitioned = AdvancedSettingsValidator.partitionSettings(settings)
        for rejection in partitioned.rejected {
            result.skippedSettings.append(
                .init(name: rejection.setting.syncKey, reason: rejection.error.message)
            )
        }

        for setting in partitioned.allowed {
            let name = setting.syncKey
            guard let typedValue = setting.typedValue else {
                result.skippedSettings.append(.init(name: name, reason: "Value could not be converted."))
                continue
            }

            do {
                try await executor.runDQL(
                    AdvancedSettingsDQL.settingStatement(for: setting),
                    arguments: ["value": typedValue.argumentValue]
                )
                result.appliedSettings.append(name)
            } catch {
                result.skippedSettings.append(.init(name: name, reason: error.localizedDescription))
            }
        }

        if result.appliedSettings.isEmpty, result.skippedSettings.isEmpty {
            Log.info("[Advanced] No startup settings to apply")
        } else {
            Log.info(
                "[Advanced] Startup settings applied=\(result.appliedSettings.count) " +
                    "skipped=\(result.skippedSettings.count)"
            )
            for skipped in result.skippedSettings {
                Log.warning("[Advanced] Skipped startup setting '\(skipped.name)': \(skipped.reason)")
            }
        }
        return result
    }

    // MARK: - Sync Scopes (fail-closed)

    /// Applies the sync scopes and verifies them by reading the parameter back.
    ///
    /// - Returns: the number of collections scoped.
    /// - Throws: if the statement fails, the scopes are invalid, or the read-back
    ///   disagrees with what was requested. The caller must not start sync on throw.
    /// - Returns: how many collections were scoped, and whether the read-back could
    ///   confirm them.
    @discardableResult
    func applySyncScopes(_ scopes: [CollectionSyncScope]) async throws -> (applied: Int, verified: Bool) {
        let map: [String: String]
        do {
            map = try AdvancedSettingsDQL.scopeMap(from: scopes)
        } catch {
            throw ApplyError.invalidScopes(String(describing: error))
        }

        // NOTE: an empty map still issues the statement. `ALTER SYSTEM` state is
        // in-memory and lives as long as the instance, so returning early on "no scopes"
        // meant deleting the last scope row never took effect — the SDK kept enforcing
        // the old `LocalPeerOnly` while the UI showed none, and the re-apply reported
        // success having sent nothing.
        do {
            try await executor.runDQL(
                AdvancedSettingsDQL.setSyncScopesQuery,
                arguments: ["scopes": map]
            )
        } catch {
            throw ApplyError.scopeStatementFailed(
                collections: map.keys.sorted(),
                underlying: error.localizedDescription
            )
        }

        // Verify. The SDK accepts the statement silently, so a read-back is the only
        // proof the containment the user asked for is actually in effect.
        switch await readSyncScopes() {
        case let .parsed(actual):
            // Subset, not equality: the instance may legitimately carry scopes this
            // config did not set (e.g. one applied by a query the user ran), and
            // demanding an exact match would refuse to open the database over it.
            for (collection, expected) in map {
                guard let found = actual[collection] else {
                    throw ApplyError.scopeVerificationMismatch(expected: map, actual: actual)
                }
                guard found == expected else {
                    throw ApplyError.scopeVerificationMismatch(expected: map, actual: actual)
                }
            }
            // Clearing case: nothing was requested, so nothing of ours may remain. A
            // leftover scope here means some other writer set it (a query the user ran),
            // which we report rather than treat as our own success.
            if map.isEmpty, !actual.isEmpty {
                Log.warning(
                    "[Advanced] Requested no sync scopes but the instance still reports " +
                        "\(actual.count): \(actual.keys.sorted().joined(separator: ", "))"
                )
                return (0, false)
            }
            Log.info("[Advanced] Sync scopes verified for \(map.count) collection(s)")
            return (map.count, true)

        case .unavailable:
            // The write succeeded; only the read-back is unusable (the SDK's `SHOW`
            // result shape for this parameter is not pinned by a live test yet). Don't
            // block the open — but never report this as verified either.
            Log.warning(
                "[Advanced] Applied \(map.count) sync scope(s) but could NOT verify them via " +
                    "'\(AdvancedSettingsDQL.showSyncScopesQuery)'. Treating as unverified."
            )
            return (map.count, false)
        }
    }

    private enum ScopeReadback {
        case parsed([String: String])
        case unavailable
    }

    /// Reads the current scope map.
    ///
    /// Deliberately does **not** fall back to `row.values.first`: on an unordered
    /// dictionary that picks an arbitrary element, which made verification succeed or
    /// fail depending on the process's hash seed — including intermittently refusing to
    /// open the database. Only known keys are consulted, and the `SHOW` statement itself
    /// is allowed to fail (some SDK builds may not support it) without bricking every
    /// database that uses scopes.
    private func readSyncScopes() async -> ScopeReadback {
        let rows: [[String: Any?]]
        do {
            rows = try await executor.runDQL(AdvancedSettingsDQL.showSyncScopesQuery, arguments: [:])
        } catch {
            Log.warning("[Advanced] Sync scope read-back query failed: \(error.localizedDescription)")
            return .unavailable
        }

        for row in rows {
            // Shape A: keyed by the parameter name (any case).
            for (key, value) in row
                where key.caseInsensitiveCompare(AdvancedSettingsDQL.syncScopesReadParameter) == .orderedSame
            {
                if let map = Self.coerceScopeMap(value) {
                    return .parsed(map)
                }
            }
            // Shape B: a name/value row, e.g. {"name": "user_collection_sync_scopes", "value": {...}}.
            if let name = row["name"].flatMap(\.self) as? String,
               name.caseInsensitiveCompare(AdvancedSettingsDQL.syncScopesReadParameter) == .orderedSame,
               let value = row["value"].flatMap(\.self),
               let map = Self.coerceScopeMap(value)
            {
                return .parsed(map)
            }
        }
        return .unavailable
    }

    /// Accepts a dictionary of strings, a loosely-typed dictionary, or a JSON string.
    private static func coerceScopeMap(_ raw: Any?) -> [String: String]? {
        guard let raw else { return nil }
        if let typed = raw as? [String: String] {
            return typed
        }
        if let loose = raw as? [String: Any] {
            var map: [String: String] = [:]
            for (key, value) in loose {
                guard let string = value as? String else { return nil }
                map[key] = string
            }
            return map
        }
        if let json = raw as? String,
           let data = json.data(using: .utf8),
           let object = try? JSONSerialization.jsonObject(with: data) as? [String: Any]
        {
            var map: [String: String] = [:]
            for (key, value) in object {
                guard let string = value as? String else { return nil }
                map[key] = string
            }
            return map
        }
        return nil
    }

    // MARK: - Reset

    /// The statement sequence a database open must perform, in order.
    ///
    /// Extracted from `DittoManager.hydrateDittoSelectedDatabase` so a test can observe
    /// the **real** ordering. The previous ordering test hand-wrote the sequence in its
    /// own body and asserted the order of its own lines, which meant moving
    /// `applySyncScopes` after `sync.start()` in production would have leaked
    /// `LocalPeerOnly` data with every test still green.
    ///
    /// Two invariants this type exists to guarantee:
    /// 1. The user's startup settings run **before** the app's own `ALTER SYSTEM`
    ///    statements, so app-managed parameters win.
    /// 2. Sync scopes are applied **and** `startSync` is never reached if they fail —
    ///    the SDK requires scopes before `start_sync()`, and a missing `LocalPeerOnly`
    ///    means data the user marked device-local replicates.
    struct OpenSequence: Sendable {
        let applier: AdvancedSettingsApplier
        /// Applies the peer-to-peer transport configuration (an SDK call, not DQL).
        let applyTransportConfig: @Sendable () async throws -> Void
        let isStrictModeEnabled: Bool
        /// The macOS-only mesh client cap; nil elsewhere.
        let meshMaxWlanClients: Int?
        /// Runs after every `ALTER SYSTEM` statement and before sync starts — the
        /// transport readback logging lives here so the log reflects final state.
        let beforeSync: @Sendable () async -> Void
        let startSync: @Sendable () async throws -> Void

        @discardableResult
        func run(
            startupSettings: [StartupSetting],
            syncScopes: [CollectionSyncScope]
        ) async throws -> AdvancedApplyResult {
            // 1. User settings first, so anything Edge Studio manages overrides them.
            var result = await applier.applyStartupSettings(startupSettings)

            // 2. Transports.
            try await applyTransportConfig()

            // 3-4. App-managed parameters.
            try await applier.executor.runDQL(
                "ALTER SYSTEM SET DQL_STRICT_MODE = \(isStrictModeEnabled ? "true" : "false")",
                arguments: [:]
            )
            if let meshMaxWlanClients {
                try await applier.executor.runDQL(
                    "ALTER SYSTEM SET mesh_chooser_max_wlan_clients = \(meshMaxWlanClients)",
                    arguments: [:]
                )
            }

            // 5. Sync scopes — fail-closed, so a throw here means `startSync` is never
            //    reached and the caller aborts the open.
            let scopeOutcome = try await applier.applySyncScopes(syncScopes)
            result.appliedScopeCount = scopeOutcome.applied
            result.scopesUnverified = !scopeOutcome.verified

            await beforeSync()

            // 6. Only now may sync start.
            try await startSync()
            return result
        }
    }

    /// Restores every system parameter to its SDK default.
    ///
    /// Callers must re-apply everything the app manages afterwards — transport
    /// configuration, `DQL_STRICT_MODE`, the macOS mesh setting and the sync scopes —
    /// because this resets *all* parameters, not just the user's.
    func resetAllToDefaults() async throws {
        try await executor.runDQL(AdvancedSettingsDQL.resetAllQuery, arguments: [:])
        Log.info("[Advanced] ALTER SYSTEM RESET ALL issued")
    }
}

File: android/app/src/main/java/com/costoda/dittoedgestudio/data/ditto/AdvancedSettingsApplier.kt
package com.costoda.dittoedgestudio.data.ditto

import android.util.Log
import com.costoda.dittoedgestudio.BuildConfig
import com.costoda.dittoedgestudio.data.repository.parseJsonToMap
import com.costoda.dittoedgestudio.domain.model.AdvancedApplyResult
import com.costoda.dittoedgestudio.domain.model.AdvancedSettingsDql
import com.costoda.dittoedgestudio.domain.model.AdvancedSettingsValidator
import com.costoda.dittoedgestudio.domain.model.CollectionSyncScope
import com.costoda.dittoedgestudio.domain.model.StartupSetting
import com.ditto.kotlin.Ditto
import org.json.JSONObject

/**
 * Minimal DQL execution surface, so the advanced-settings apply logic can be tested
 * with a recording fake instead of a live Ditto instance.
 */
interface DQLExecuting {
    /** Runs [query] and returns each result item's value map. */
    suspend fun runDQL(query: String, arguments: Map<String, Any?>): List<Map<String, Any?>>
}

class DittoDQLExecutor(private val ditto: Ditto) : DQLExecuting {
    override suspend fun runDQL(
        query: String,
        arguments: Map<String, Any?>,
    ): List<Map<String, Any?>> = ditto.store.execute(query, arguments) { result ->
        result.items.map { item ->
            runCatching { parseJsonToMap(JSONObject(item.jsonString())) }
                .getOrDefault(emptyMap())
        }
    }
}

/**
 * Applies a database's Advanced Configuration to an open Ditto instance.
 *
 * Split out of [DittoManager] for two reasons: the statement order relative to
 * `sync.start()` is safety-critical and needs a unit test that doesn't require
 * credentials, and the two lists need **opposite failure policies**:
 *
 * - **Startup settings** are tuning knobs → best-effort. One bad parameter name is
 *   reported, not fatal.
 * - **Sync scopes** are a data-containment control → fail-closed. If they cannot be
 *   applied and verified, the caller must not start sync, because "scope missing"
 *   means data the user marked device-local would replicate.
 *
 * Port of SwiftUI `AdvancedSettingsApplier` — see docs/ADVANCED_DATABASE_CONFIG.md.
 */
class AdvancedSettingsApplier(private val executor: DQLExecuting) {

    sealed class ApplyError(message: String) : Exception(message) {
        data class ScopeStatementFailed(val collections: List<String>, val underlying: String) :
            ApplyError(
                "Could not apply collection sync scopes for ${collections.joinToString(", ")}. " +
                    "The database was not opened, to avoid syncing data you marked device-local. " +
                    "($underlying)",
            )

        data class ScopeVerificationMismatch(
            val expected: Map<String, String>,
            val actual: Map<String, String>,
        ) : ApplyError(
            "Collection sync scopes were not stored as requested " +
                "(expected $expected, found $actual). The database was not opened, " +
                "to avoid syncing data you marked device-local.",
        )

        data class InvalidScopes(val detail: String) :
            ApplyError("Collection sync scopes are invalid: $detail. The database was not opened.")
    }

    // MARK: - Startup Settings (best-effort)

    /**
     * Applies each startup setting independently. A failure is recorded and skipped
     * rather than aborting: an unrecognized parameter name is a typo, not a reason to
     * lock the user out of their database.
     */
    suspend fun applyStartupSettings(settings: List<StartupSetting>): AdvancedApplyResult {
        var applied = listOf<String>()
        var skipped = listOf<AdvancedApplyResult.Skipped>()
        if (settings.isEmpty()) return AdvancedApplyResult()

        // Whole-list validation, not per-row: `validateSetting(setting, others = [])`
        // cannot see duplicates, and the row cap was only ever enforced in the editor —
        // so a hand-edited or imported config could issue thousands of statements, or
        // two conflicting writes to one parameter, silently.
        val (allowed, rejected) = AdvancedSettingsValidator.partitionSettings(settings)
        skipped = skipped + rejected.map { (setting, error) ->
            AdvancedApplyResult.Skipped(setting.syncKey, error.message)
        }

        for (setting in allowed) {
            val name = setting.syncKey
            val typedValue = setting.typedValue()
            if (typedValue == null) {
                skipped = skipped + AdvancedApplyResult.Skipped(name, "Value could not be converted.")
                continue
            }
            try {
                executor.runDQL(
                    AdvancedSettingsDql.settingStatement(setting),
                    mapOf("value" to typedValue.argumentValue()),
                )
                applied = applied + name
            } catch (e: Exception) {
                skipped = skipped + AdvancedApplyResult.Skipped(name, e.message ?: e.toString())
            }
        }

        if (BuildConfig.DEBUG) {
            if (applied.isEmpty() && skipped.isEmpty()) {
                Log.i(TAG, "[Advanced] No startup settings to apply")
            } else {
                Log.i(TAG, "[Advanced] Startup settings applied=${applied.size} skipped=${skipped.size}")
                skipped.forEach { Log.w(TAG, "[Advanced] Skipped startup setting '${it.name}': ${it.reason}") }
            }
        }
        return AdvancedApplyResult(appliedSettings = applied, skippedSettings = skipped)
    }

    // MARK: - Sync Scopes (fail-closed)

    /**
     * Applies the sync scopes and verifies them by reading the parameter back.
     *
     * @return how many collections were scoped, and whether the read-back confirmed them.
     * @throws ApplyError if the statement fails, the scopes are invalid, or the read-back
     *   disagrees with what was requested. The caller must not start sync on throw.
     */
    suspend fun applySyncScopes(scopes: List<CollectionSyncScope>): Pair<Int, Boolean> {
        val map = try {
            AdvancedSettingsDql.scopeMap(scopes)
        } catch (e: Exception) {
            throw ApplyError.InvalidScopes(e.message ?: e.toString())
        }

        // NOTE: an empty map still issues the statement. `ALTER SYSTEM` state is
        // in-memory and lives as long as the instance, so returning early on "no scopes"
        // meant deleting the last scope row never took effect — the SDK kept enforcing
        // the old `LocalPeerOnly` while the UI showed none, and the re-apply reported
        // success having sent nothing.
        try {
            executor.runDQL(AdvancedSettingsDql.SET_SYNC_SCOPES_QUERY, mapOf("scopes" to map))
        } catch (e: Exception) {
            throw ApplyError.ScopeStatementFailed(
                collections = map.keys.sorted(),
                underlying = e.message ?: e.toString(),
            )
        }

        // Verify. The SDK accepts the statement silently, so a read-back is the only
        // proof the containment the user asked for is actually in effect.
        return when (val readback = readSyncScopes()) {
            is ScopeReadback.Parsed -> {
                val actual = readback.map
                // Subset, not equality: the instance may legitimately carry scopes this
                // config did not set (e.g. one applied by a query the user ran), and
                // demanding an exact match would refuse to open the database over it.
                for ((collection, expected) in map) {
                    val found = actual[collection]
                    if (found == null || found != expected) {
                        throw ApplyError.ScopeVerificationMismatch(expected = map, actual = actual)
                    }
                }
                // Clearing case: nothing was requested, so nothing of ours may remain. A
                // leftover scope here means some other writer set it (a query the user ran),
                // which we report rather than treat as our own success.
                if (map.isEmpty() && actual.isNotEmpty()) {
                    if (BuildConfig.DEBUG) {
                        Log.w(
                            TAG,
                            "[Advanced] Requested no sync scopes but the instance still reports " +
                                "${actual.size}: ${actual.keys.sorted().joinToString(", ")}",
                        )
                    }
                    return 0 to false
                }
                if (BuildConfig.DEBUG) {
                    Log.i(TAG, "[Advanced] Sync scopes verified for ${map.size} collection(s)")
                }
                map.size to true
            }

            ScopeReadback.Unavailable -> {
                // The write succeeded; only the read-back is unusable. Don't block the
                // open — but never report this as verified either.
                if (BuildConfig.DEBUG) {
                    Log.w(
                        TAG,
                        "[Advanced] Applied ${map.size} sync scope(s) but could NOT verify them via " +
                            "'${AdvancedSettingsDql.SHOW_SYNC_SCOPES_QUERY}'. Treating as unverified.",
                    )
                }
                map.size to false
            }
        }
    }

    private sealed interface ScopeReadback {
        data class Parsed(val map: Map<String, String>) : ScopeReadback
        data object Unavailable : ScopeReadback
    }

    /**
     * Reads the current scope map.
     *
     * Deliberately does **not** fall back to "the first value in the row": on an
     * unordered map that picks an arbitrary element, making verification succeed or
     * fail unpredictably. Only known keys are consulted, and the `SHOW` statement
     * itself is allowed to fail without bricking every database that uses scopes.
     */
    private suspend fun readSyncScopes(): ScopeReadback {
        val rows = try {
            executor.runDQL(AdvancedSettingsDql.SHOW_SYNC_SCOPES_QUERY, emptyMap())
        } catch (e: Exception) {
            if (BuildConfig.DEBUG) {
                Log.w(TAG, "[Advanced] Sync scope read-back query failed: ${e.message}")
            }
            return ScopeReadback.Unavailable
        }

        for (row in rows) {
            // Shape A: keyed by the parameter name (any case).
            for ((key, value) in row) {
                if (key.equals(AdvancedSettingsDql.SYNC_SCOPES_READ_PARAMETER, ignoreCase = true)) {
                    coerceScopeMap(value)?.let { return ScopeReadback.Parsed(it) }
                }
            }
            // Shape B: a name/value row, e.g. {"name": "user_collection_sync_scopes", "value": {...}}.
            val name = row["name"] as? String
            if (name != null && name.equals(AdvancedSettingsDql.SYNC_SCOPES_READ_PARAMETER, ignoreCase = true)) {
                coerceScopeMap(row["value"])?.let { return ScopeReadback.Parsed(it) }
            }
        }
        return ScopeReadback.Unavailable
    }

    /** Accepts a map of strings or a JSON object string; anything else is unreadable. */
    private fun coerceScopeMap(raw: Any?): Map<String, String>? = when (raw) {
        null -> null
        is Map<*, *> -> {
            val out = mutableMapOf<String, String>()
            for ((key, value) in raw) {
                if (key !is String || value !is String) return null
                out[key] = value
            }
            out
        }
        is String -> runCatching {
            val obj = JSONObject(raw)
            val out = mutableMapOf<String, String>()
            for (key in obj.keys()) {
                val value = obj.get(key)
                if (value !is String) return null
                out[key] = value
            }
            out
        }.getOrNull()
        else -> null
    }

    // MARK: - Open Sequence

    /**
     * The statement sequence a database open must perform, in order.
     *
     * Extracted from [DittoManager.hydrate] so a test can observe the **real**
     * ordering. Two invariants this type exists to guarantee:
     * 1. The user's startup settings run **before** the app's own `ALTER SYSTEM`
     *    statements, so app-managed parameters win.
     * 2. Sync scopes are applied **and** [startSync] is never reached if they fail —
     *    the SDK requires scopes before `start_sync()`, and a missing `LocalPeerOnly`
     *    means data the user marked device-local replicates.
     *
     * (`mesh_chooser_max_wlan_clients` from the SwiftUI sequence is macOS-only and has
     * no Android equivalent.)
     */
    class OpenSequence(
        private val applier: AdvancedSettingsApplier,
        /** Applies the peer-to-peer transport configuration (an SDK call, not DQL). */
        private val applyTransportConfig: suspend () -> Unit,
        private val isStrictModeEnabled: Boolean,
        private val startSync: suspend () -> Unit,
    ) {
        suspend fun run(
            startupSettings: List<StartupSetting>,
            syncScopes: List<CollectionSyncScope>,
        ): AdvancedApplyResult {
            // 1. User settings first, so anything Edge Studio manages overrides them.
            var result = applier.applyStartupSettings(startupSettings)

            // 2. Transports.
            applyTransportConfig()

            // 3. App-managed parameters.
            applier.executor.runDQL(
                "ALTER SYSTEM SET DQL_STRICT_MODE = ${if (isStrictModeEnabled) "true" else "false"}",
                emptyMap(),
            )

            // 4. Sync scopes — fail-closed, so a throw here means startSync is never
            //    reached and the caller aborts the open.
            val (appliedScopes, verified) = applier.applySyncScopes(syncScopes)
            result = result.copy(appliedScopeCount = appliedScopes, scopesUnverified = !verified)

            // 5. Only now may sync start.
            startSync()
            return result
        }
    }

    /**
     * Restores every system parameter to its SDK default.
     *
     * Callers must re-apply everything the app manages afterwards — transport
     * configuration, `DQL_STRICT_MODE` and the sync scopes — because this resets
     * *all* parameters, not just the user's.
     */
    suspend fun resetAllToDefaults() {
        executor.runDQL(AdvancedSettingsDql.RESET_ALL_QUERY, emptyMap())
        if (BuildConfig.DEBUG) Log.i(TAG, "[Advanced] ALTER SYSTEM RESET ALL issued")
    }

    private companion object {
        const val TAG = "AdvancedSettings"
    }
}
